diff --git a/nptl/sysdeps/pthread/pt-initfini.c b/nptl/sysdeps/pthread/pt-initfini.c
index 9c00dc0b7dc058e4fa..5949f7b7a3b5025988 100644
--- a/nptl/sysdeps/pthread/pt-initfini.c
+++ b/nptl/sysdeps/pthread/pt-initfini.c
@@ -44,6 +44,10 @@
 
 /* Embed an #include to pull in the alignment and .end directives. */
 asm ("\n#include \"defs.h\"");
+asm ("\n#ifdef __i686");
+asm ("\n# undef __i686");
+asm ("\n# define __i686 __i686");
+asm ("\n#endif");
 
 /* The initial common code ends here. */
 asm ("\n/*@HEADER_ENDS*/");
diff --git a/nptl/sysdeps/unix/sysv/linux/i386/i486/pthread_cond_wait.S b/nptl/sysdeps/unix/sysv/linux/i386/i486/pthread_cond_wait.S
index 5b301979b534fdbb45..730f3588649f45fe2b 100644
--- a/nptl/sysdeps/unix/sysv/linux/i386/i486/pthread_cond_wait.S
+++ b/nptl/sysdeps/unix/sysv/linux/i386/i486/pthread_cond_wait.S
@@ -24,6 +24,10 @@
 #include <tcb-offsets.h>
 #include <kernel-features.h>
 
+#if defined(PIC) && defined(__i686)
+# undef __i686
+# define __i686 __i686
+#endif
 
 	.text
 
diff --git a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_post.S b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_post.S
index 86992c877e141df39f..6fb171ee934a69e3b6 100644
--- a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_post.S
+++ b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_post.S
@@ -23,6 +23,10 @@
 #include <structsem.h>
 #include <lowlevellock.h>
 
+#if defined(PIC) && defined(__i686)
+# undef __i686	.text
+# define __i686 __i686
+#endif	.globl	__new_sem_post
 
 	.text
 
diff --git a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_timedwait.S b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_timedwait.S
index 13a36d64cc7dd61147..7f63e261e5a1b1bea0 100644
--- a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_timedwait.S
+++ b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_timedwait.S
@@ -23,6 +23,10 @@
 #include <structsem.h>
 #include <lowlevellock.h>
 
+#if defined(PIC) && defined(__i686)
+# undef __i686
+# define __i686 __i686
+#endif
 
 #if VALUE != 0
 # error "code needs to be rewritten for VALUE != 0"
diff --git a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_trywait.S b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_trywait.S
index 2d49934f02c1eb2d37..aba68500e5e2e3fa21 100644
--- a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_trywait.S
+++ b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_trywait.S
@@ -22,6 +22,11 @@
 #include <pthread-errnos.h>
 #include <lowlevellock.h>
 
+#if defined(PIC) && defined(__i686)
+# undef __i686
+# define __i686 __i686
+#endif
+
 	.text
 
 	.globl	__new_sem_trywait
diff --git a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_wait.S b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_wait.S
index 93c0a64a9e47299966..32187ddf5a6b2330b5 100644
--- a/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_wait.S
+++ b/nptl/sysdeps/unix/sysv/linux/i386/i486/sem_wait.S
@@ -23,6 +23,10 @@
 #include <structsem.h>
 #include <lowlevellock.h>
 
+#if defined(PIC) && defined(__i686)
+# undef __i686
+# define __i686 __i686
+#endif
 
 #if VALUE != 0
 # error "code needs to be rewritten for VALUE != 0"
diff --git a/nptl/sysdeps/unix/sysv/linux/i386/pthread_once.S b/nptl/sysdeps/unix/sysv/linux/i386/pthread_once.S
index f31d968bc5a4d24984..796fdeb4ccca7a3e0d 100644
--- a/nptl/sysdeps/unix/sysv/linux/i386/pthread_once.S
+++ b/nptl/sysdeps/unix/sysv/linux/i386/pthread_once.S
@@ -22,6 +22,10 @@
 #include <kernel-features.h>
 #include <lowlevellock.h>
 
+#if defined(PIC) && defined(__i686)
+# undef __i686
+# define __i686 __i686
+#endif
 
 	.comm	__fork_generation, 4, 4
 
diff --git a/sysdeps/i386/i686/memcmp.S b/sysdeps/i386/i686/memcmp.S
index 24f6804b0c17e1f1d4..5081f60a72e5438aa4 100644
--- a/sysdeps/i386/i686/memcmp.S
+++ b/sysdeps/i386/i686/memcmp.S
@@ -22,6 +22,11 @@
 #include "bp-sym.h"
 #include "bp-asm.h"
 
+#ifdef __i686
+# undef __i686
+# define __i686 __i686
+#endif
+
 #define PARMS		LINKAGE+4	/* Preserve EBX.  */
 #define BLK1		PARMS
 #define BLK2		BLK1+PTR_SIZE
diff --git a/sysdeps/unix/sysv/linux/i386/sysdep.h b/sysdeps/unix/sysv/linux/i386/sysdep.h
index 3b7acd4bee71ea7484..da9991299668dbe525 100644
--- a/sysdeps/unix/sysv/linux/i386/sysdep.h
+++ b/sysdeps/unix/sysv/linux/i386/sysdep.h
@@ -29,7 +29,6 @@
 #include <dl-sysdep.h>
 #include <tls.h>
 
-
 /* For Linux we can use the system call table in the header file
 	/usr/include/asm/unistd.h
    of the kernel.  But these symbols do not follow the SYS_* syntax
@@ -535,6 +534,11 @@ asm (".L__X'%ebx = 1\n\t"
 	".previous"							      \
 	: "=c" (__res));						      \
      __res; })
+#ifdef __i686
+# undef __i686
+# define __i686 __i686
+#endif
+
 #endif
 #endif
 
