diff --git a/elf/Makefile b/elf/Makefile
index 85f8d09781773e28c8..d8c5ab888bc6b2042e 100644
--- a/elf/Makefile
+++ b/elf/Makefile
@@ -315,7 +315,7 @@ $(objpfx)ld.so: $(objpfx)librtld.os $(ld-map)
 		  $(LDFLAGS-rtld) -Wl,-z,defs -Wl,--verbose 2>&1 |	\
 		  LC_ALL=C \
 		  sed -e '/^=========/,/^=========/!d;/^=========/d' \
-		      -e 's/\. = 0 + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
+		      -e 's/\. = .* + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
 		      -e 's/_end = \.; PROVIDE (end = \.);/& . = DATA_SEGMENT_END(.);/' \
 		      -e 's/seg_bss      PT_LOAD FLAGS(6) ;       \/\* read + write \*\///' \
 		      -e 's/} :seg_bss/}/' \
diff --git a/manual/Makefile b/manual/Makefile
index c5866eb9def3becfc6..f82c9efe188540e8df 100644
--- a/manual/Makefile
+++ b/manual/Makefile
@@ -232,7 +232,8 @@ ifdef objpfx
 .PHONY: stubs
 stubs: $(objpfx)stubs
 endif
-$(objpfx)stubs ../po/manual.pot $(objpfx)stamp%:
+stamps := $(foreach o,$(object-suffixes-for-libc),$(objpfx)stamp$o)
+$(objpfx)stubs ../po/manual.pot $(objpfx)stamp.o $(stamps):
 	$(make-target-directory)
 	touch $@
 
@@ -245,7 +246,7 @@ endef
 
 # The top-level glibc Makefile expects subdir_install to update the stubs file.
 subdir_install: stubs
-
+
 # Get rid of these variables if they came from the parent.
 routines =
 aux =
