diff --git a/misc/Makefile b/misc/Makefile
index eb97ac17674cfdbe64..351d87edaa7f47df73 100644
--- a/misc/Makefile
+++ b/misc/Makefile
@@ -64,7 +64,7 @@ routines := brk sbrk sstk ioctl \
 	    getloadavg getclktck \
 	    fgetxattr flistxattr fremovexattr fsetxattr getxattr \
 	    listxattr lgetxattr llistxattr lremovexattr lsetxattr \
-	    removexattr setxattr
+	    removexattr setxattr unimpl_syscall
 
 distribute := device-nrs.h
 
@@ -81,7 +81,7 @@ gpl2lgpl := error.c error.h
 tests := tst-dirname tst-tsearch tst-fdset tst-efgcvt tst-mntent tst-hsearch \
 	 tst-error1 tst-pselect tst-insremque tst-mntent2
 ifeq (no,$(cross-compiling))
-tests: $(objpfx)tst-error1-mem
+tests: $(objpfx)tst-error1-memren
 endif
 
 CFLAGS-tsearch.c = $(uses-callbacks)
diff --git a/nptl/Makefile b/nptl/Makefile
index aaa9685c322ece093d..e6b237ba226a49a4e4 100644
--- a/nptl/Makefile
+++ b/nptl/Makefile
@@ -125,7 +125,7 @@ libpthread-routines = init vars events version \
 		      pthread_mutexattr_getprioceiling \
 		      pthread_mutexattr_setprioceiling tpp \
 		      pthread_mutex_getprioceiling pthread_mutex_setprioceiling \
-		      lind_rpc lind_syscalls lind_rpc_gen lind_util lindlock lind_file
+		      imc_sendmsg imc_recvmsg lind_rpc lind_syscalls lind_rpc_gen nacl_util lindlock
 #		      pthread_setuid pthread_seteuid pthread_setreuid \
 #		      pthread_setresuid \
 #		      pthread_setgid pthread_setegid pthread_setregid \
diff --git a/sysdeps/i386/Makefile b/sysdeps/i386/Makefile
index efeb5be98a50d4fb1e..536cc529617f6ac884 100644
--- a/sysdeps/i386/Makefile
+++ b/sysdeps/i386/Makefile
@@ -60,7 +60,6 @@ CFLAGS-dl-fini.c += -mpreferred-stack-boundary=4
 CFLAGS-dl-open.c += -mpreferred-stack-boundary=4
 CFLAGS-dl-close.c += -mpreferred-stack-boundary=4
 CFLAGS-dl-error.c += -mpreferred-stack-boundary=4
-CFLAGS-rtld.c += -mpreferred-stack-boundary=4
 endif
 ifeq ($(subdir),dlfcn)
 CFLAGS-dlopen.c += -mpreferred-stack-boundary=4
diff --git a/sysdeps/nacl/Makefile b/sysdeps/nacl/Makefile
index 212f528f192a061be0..a1cb02eeac20a129af 100644
--- a/sysdeps/nacl/Makefile
+++ b/sysdeps/nacl/Makefile
@@ -1,13 +1,14 @@
+
 ifeq ($(subdir),io)
-sysdep_routines += lind_rpc lind_util lind_syscalls lind_rpc_gen lindlock lind_file
+sysdep_routines += imc_sendmsg imc_recvmsg imc_accept lind_rpc nacl_util lind_syscalls lind_rpc_gen lindlock
 endif
 
-rtld-installed-name = $(patsubst %.$(NACL_FAKE_SONAME),%.1,$(ld.so-version))
+# ifeq ($(subdir),socket)
+# sysdep_routines += socket
+# endif
+
+
 
 # Remove libpthread.so's version of a function that is also provided
 # by libc.so.
 libpthread-routines := $(filter-out pt-vfork,$(libpthread-routines))
-
-# In case of NaCl any program which wants to call syscalls must be linked with
-# ld.so explicitly.
-LDLIBS-nscd += $(elfobjdir)/ld.so
