diff --git a/elf/Makefile b/elf/Makefile
index 8a3f5b7905514c291d..d5aa7e0f6d1559c15b 100644
--- a/elf/Makefile
+++ b/elf/Makefile
@@ -23,8 +23,7 @@ subdir		:= elf
 headers		= elf.h bits/elfclass.h link.h bits/link.h
 routines	= $(dl-routines) dl-support dl-iteratephdr \
 		  dl-addr enbl-secure dl-profstub \
-		  dl-origin dl-libc dl-sym dl-tsd dl-sysdep \
-		  receive_args
+		  dl-origin dl-libc dl-sym dl-tsd dl-sysdep
 
 # The core dynamic linking functions are in libc for the static and
 # profiled libraries.
@@ -32,8 +31,8 @@ dl-routines	= $(addprefix dl-,load cache lookup object reloc deps \
 			          runtime error init fini debug misc \
 				  version profile conflict tls origin \
 				  execstack caller open close trampoline) \
-		  nacl_tls_init nacl_read_tp \
-		  nacl_dyncode_copy nacl_dyncode_map nacl_dyncode_alloc \
+		  nacl_read_tp irt_syscalls \
+		  nacl_dyncode_create nacl_dyncode_map nacl_dyncode_alloc \
 		  futex_emulation nacl_dyncode_valgrind
 all-dl-routines = $(dl-routines) $(sysdep-dl-routines)
 # But they are absent from the shared libc, because that code is in ld.so.
@@ -304,13 +303,26 @@ generated += librtld.map librtld.mk rtld-libc.a librtld.os.map
 
 z-now-yes = -Wl,-z,now
 
+
+
 $(objpfx)ld.so: $(objpfx)librtld.os $(ld-map)
 	@rm -f $@.lds
+	# Modify elf64_nacl.xs and elf_nacl.xs linker scripts for ld.so.
+	# NaCl loader refuses to run ld.so with more than 3 loadable segments.
+	# Remove the bss segment.
 	$(LINK.o) -nostdlib -nostartfiles -shared $(z-now-$(bind-now))	\
 		  $(LDFLAGS-rtld) -Wl,-z,defs -Wl,--verbose 2>&1 |	\
 		  LC_ALL=C \
-		  sed -e '/^=========/,/^=========/!d;/^=========/d'	\
+		  sed -e '/^=========/,/^=========/!d;/^=========/d' \
 		      -e 's/\. = 0 + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
+		      -e 's/_end = \.; PROVIDE (end = \.);/& . = DATA_SEGMENT_END(.);/' \
+		      -e 's/seg_bss      PT_LOAD FLAGS(6) ;       \/\* read + write \*\///' \
+		      -e 's/} :seg_bss/}/' \
+		      -e 's/\. = DATA_SEGMENT_END (\.);//' \
+		      -e 's/\. = SEGMENT_START("text", 0) + SIZEOF_HEADERS;/\. = SEGMENT_START("text", 0);/' \
+		      -e 's/(\. & (CONSTANT (MAXPAGESIZE) - 1));/0;/' \
+		      -e 's/\. = ALIGN (CONSTANT (MAXPAGESIZE));//' \
+		      -e 's/\. = ALIGN (CONSTANT (COMMONPAGESIZE));//' \
 		  > $@.lds
 	$(LINK.o) -nostdlib -nostartfiles -shared -o $@			\
 		  $(LDFLAGS-rtld) -Wl,-z,defs $(z-now-$(bind-now))	\
@@ -353,7 +365,7 @@ $(objpfx)trusted-dirs.st: Makefile $(..)Makeconfig
 	$(make-target-directory)
 	echo "$(subst :, ,$(default-rpath) $(user-defined-trusted-dirs))"    \
 	| $(AWK) -f gen-trusted-dirs.awk > ${@:st=T};
-	echo '#define DL_DST_LIB "$(notdir $(slibdir))"' >> ${@:st=T}
+	echo '#define DL_DST_LIB "$(default-dl-dst-lib)"' >> ${@:st=T}
 	$(move-if-change) ${@:st=T} ${@:st=h}
 	touch $@
 CPPFLAGS-dl-load.c = -I$(objpfx). -I$(csu-objpfx).
diff --git a/manual/Makefile b/manual/Makefile
index c5866eb9def3becfc6..f82c9efe188540e8df 100644
--- a/manual/Makefile
+++ b/manual/Makefile
@@ -232,7 +232,8 @@ ifdef objpfx
 .PHONY: stubs
 stubs: $(objpfx)stubs
 endif
-$(objpfx)stubs ../po/manual.pot $(objpfx)stamp%:
+stamps := $(foreach o,$(object-suffixes-for-libc),$(objpfx)stamp$o)
+$(objpfx)stubs ../po/manual.pot $(objpfx)stamp.o $(stamps):
 	$(make-target-directory)
 	touch $@
 
