diff --git a/build/Makefile b/build/Makefile
deleted file mode 100644
index a715be10f51395f8c0..000000000000000000
--- a/build/Makefile
+++ /dev/null
@@ -1,16 +0,0 @@
-srcdir = /home/cmatthew/lind/nacl-glibc
-
-# Uncomment the line below if you want to do parallel build.
-# PARALLELMFLAGS = -j 4
-
-# This option is for those who modify the sources and keep them in a
-# CVS repository.  Sometimes it is necessary to pass options to the cvs
-# program (not the command), like -z9 or -x.
-# CVSOPTS = -z9
-
-all .DEFAULT:
-	$(MAKE) -r PARALLELMFLAGS="$(PARALLELMFLAGS)" CVSOPTS="$(CVSOPTS)" -C $(srcdir) objdir=`pwd` $@
-
-install:
-	LANGUAGE=C LC_ALL=C; export LANGUAGE LC_ALL; \
-	$(MAKE) -r PARALLELMFLAGS="$(PARALLELMFLAGS)" CVSOPTS="$(CVSOPTS)" -C $(srcdir) objdir=`pwd` $@
