diff --git a/src/trusted/service_runtime/export_header.py b/src/trusted/service_runtime/export_header.py
index 480638317349b4de20..30d56fb454a745d978 100755
--- a/src/trusted/service_runtime/export_header.py
+++ b/src/trusted/service_runtime/export_header.py
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/bin/python2
 # Copyright (c) 2008 The Native Client Authors. All rights reserved.
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
