diff --git a/tools/Makefile b/tools/Makefile
index 3389203cc123b920fd..57a58587ece44ade33 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -43,7 +43,7 @@ SHELL = /bin/bash
 # or add to git-cl codereview.settings:
 #   PUSH_URL_CONFIG: url.ssh://gerrit.chromium.org.pushinsteadof
 #   ORIGIN_URL_CONFIG: http://git.chromium.org
-GIT_BASE_URL = http://git.chromium.org/native_client
+GIT_BASE_URL = https://chromium.googlesource.com/native_client
 
 CROSSARCH = x86_64-nacl
 TOOLCHAINLOC ?= out
@@ -338,23 +338,35 @@ sdkdirs:
 # Builds the cross assembler, linker, archiver, etc.
 ##################################################################
 BUILD/stamp-$(CROSSARCH)-binutils: | SRC/binutils BUILD
-	rm -rf BUILD/build-binutils-$(CROSSARCH)
-	mkdir BUILD/build-binutils-$(CROSSARCH)
-	# We'd like to build binutils with -Werror, but there are a
-	# number of warnings in the Mac version of GCC that prevent
-	# us from building with -Werror today.
-	cd BUILD/build-binutils-$(CROSSARCH) && \
-	  CC="$(GCC_CC)" \
-	  CFLAGS="$(USER_CFLAGS)" \
-	  LDFLAGS="$(USER_LDFLAGS)" \
-	  ../../SRC/binutils/configure \
-	    --prefix=$(PREFIX) \
-	    --target=$(CROSSARCH) \
-	    --with-sysroot=$(PREFIX)/$(CROSSARCH) \
-	    --disable-werror --enable-deterministic-archives --without-zlib
-	$(MAKE) -C BUILD/build-binutils-$(CROSSARCH) all
-	$(MAKE) -C BUILD/build-binutils-$(CROSSARCH) DESTDIR=$(DESTDIR) install
-	touch $@
+	 # We'd like to build binutils with -Werror, but there are a
+	 # number of warnings in the Mac version of GCC that prevent
+	 # us from building with -Werror today.
+	 rm -rf BUILD/build-binutils-$(CROSSARCH) && \
+	   mkdir BUILD/build-binutils-$(CROSSARCH) && \
+	   cd BUILD/build-binutils-$(CROSSARCH) && \
+	   CC="$(GCC_CC)" \
+	   CFLAGS="$(USER_CFLAGS) -D_P_NOWAITO=3" \
+	   LDFLAGS="$(USER_LDFLAGS)" \
+	   ../../SRC/binutils/configure \
+	     CFLAGS_FOR_TARGET="-O2 -g $(GCC_CFLAGS_FOR_TARGET-nolibc)" \
+	     CXXFLAGS_FOR_TARGET="-O2 -g $(GCC_CFLAGS_FOR_TARGET-nolibc)" \
+	     --prefix=$(PREFIX) \
+	     --target=$(CROSSARCH) \
+	     --with-sysroot=$(PREFIX)/$(CROSSARCH) \
+	     --disable-werror --enable-deterministic-archives --without-zlib && \
+	   $(MAKE) -C BUILD/build-binutils-$(CROSSARCH) maybe-configure-libiberty && \
+	   sed \
+	     -e '/HAVE_PSTAT_GETDYNAMIC/ s/1/0/' \
+	     -e '/HAVE_PSTAT_GETSTATIC/ s/1/0/' \
+	     <BUILD/build-binutils-$(CROSSARCH)/libiberty/config.h \
+	     >BUILD/build-binutils-$(CROSSARCH)/libiberty/config.h.new && \
+	   cat \
+	     <BUILD/build-binutils-$(CROSSARCH)/libiberty/config.h.new \
+	     >BUILD/build-binutils-$(CROSSARCH)/libiberty/config.h && \
+	   rm -f BUILD/build-binutils-$(CROSSARCH)/libiberty/config.h.new && \
+	   $(MAKE) -C BUILD/build-binutils-$(CROSSARCH) all && \
+	   $(MAKE) -C BUILD/build-binutils-$(CROSSARCH) DESTDIR=$(DESTDIR) install && \
+	   touch $@
 
 .PHONY: binutils
 binutils: BUILD/stamp-$(CROSSARCH)-binutils
