diff --git a/include/httpd.h b/include/httpd.h
index 65392f835464fd428d..14f0b941e09ae550f2 100644
--- a/include/httpd.h
+++ b/include/httpd.h
@@ -68,10 +68,12 @@
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+typedef struct apr_thread_mutex_t apr_thread_mutex_t;
+
 /* ----------------------------- config dir ------------------------------ */
 
 /** Define this to be the default server home dir. Most things later in this
  * file with a relative pathname will have this added.
  */
