diff --git a/modules/proxy/proxy_util.c b/modules/proxy/proxy_util.c
index f5ede0ff1f36172984..cefb571fbe4a0a85a7 100644
--- a/modules/proxy/proxy_util.c
+++ b/modules/proxy/proxy_util.c
@@ -12,10 +12,14 @@
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
 
+
+#include <apr_thread_mutex.h>
+#define APR_THREAD_MUTEX_DEFAULT  0x0
+
 /* Utility routines for Apache proxy */
 #include "mod_proxy.h"
 #include "ap_mpm.h"
 #include "scoreboard.h"
 #include "apr_version.h"
@@ -40,12 +44,10 @@
 #endif
 #if (APR_MAJOR_VERSION < 2)
 #include "apr_support.h"        /* for apr_wait_for_io_or_timeout() */
 #endif
 
-#include "apr_thread_mutex.h"
-
 APLOG_USE_MODULE(proxy);
 
 /*
  * Opaque structure containing target server info when
  * using a forward proxy.
diff --git a/modules/proxy/mod_proxy_balancer.c b/modules/proxy/mod_proxy_balancer.c
index c59f5e973fc7ca825f..77fa9d6c39fcc3b78a 100644
--- a/modules/proxy/mod_proxy_balancer.c
+++ b/modules/proxy/mod_proxy_balancer.c
@@ -22,10 +22,11 @@
 #include "apr_version.h"
 #include "ap_hooks.h"
 #include "apr_date.h"
 #include "util_md5.h"
 #include "mod_watchdog.h"
+#include <apr_thread_mutex.h>
 
 static const char *balancer_mutex_type = "proxy-balancer-shm";
 ap_slotmem_provider_t *storage = NULL;
 
 module AP_MODULE_DECLARE_DATA proxy_balancer_module;
