/* PR c++/30759 */
/* { dg-do "compile" } */

struct A {
   A(int) { }
};

struct B {
   B(const B&);			// { dg-message "candidate" }
   int b;
};

struct C {};

struct D { int c; };

int main()
{
   int i = { 1 };
   int j = { 1, 2 }; /* { dg-error "requires one element" } */
   A a = { 6 }; /* { dg-error "initialize" } */
   B b = { 6 }; /* { dg-error "initialize" } */
   C c = { 6 }; /* { dg-error "too many initializers" } */
   D d = { 6 };
}
