diff --git a/glob/glob.c b/glob/glob.c
index a4f76d7d312d503831..35ce63210b261751f8 100644
--- a/glob/glob.c
+++ b/glob/glob.c
@@ -253,10 +252,13 @@ extern char *alloca ();
 # ifndef __stat
 #  define __stat(fname, buf) __xstat (_STAT_VER, fname, buf)
 # endif
 #endif
 
+#define __alloca alloca
+#define __stat stat
+
 #if !(defined STDC_HEADERS || defined __GNU_LIBRARY__)
 # undef	size_t
 # define size_t	unsigned int
 #endif
 
