diff --git a/gcc/toplev.c b/gcc/toplev.c
index 5089965208f89f390b..273bbd483bdade8176 100644
--- a/gcc/toplev.c
+++ b/gcc/toplev.c
@@ -527,7 +527,7 @@ read_integral_parameter (const char *p, const char *pname, const int  defval)
    for floor_log2 and exact_log2; see toplev.h.  That construct, however,
    conflicts with the ISO C++ One Definition Rule.   */
 
-#if GCC_VERSION < 3004 || !defined (__cplusplus)
+#if GCC_VERSION < 3004 && !defined (__cplusplus)
 
 /* Given X, an unsigned number, return the largest int Y such that 2**Y <= X.
    If X is 0, return -1.  */
