diff --git a/SConstruct b/SConstruct
index 5afd2652cffa0081c9..0bc0fe1b617d07d78c 100755
--- a/SConstruct
+++ b/SConstruct
@@ -1,4 +1,4 @@
-#! -*- python -*-
+#! -*- python2 -*-
 # Copyright (c) 2012 The Native Client Authors. All rights reserved.
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
@@ -292,7 +292,7 @@ def SetUpArgumentBits(env):
     desc='Collect and display information about which commands are executed '
       'during the build process')
 
-  BitFromArgument(env, 'werror', default=True,
+  BitFromArgument(env, 'werror', default=False,
     desc='Treat warnings as errors (-Werror)')
 
   BitFromArgument(env, 'disable_nosys_linker_warnings', default=False,
@@ -2718,8 +2718,10 @@ def MakeLinuxEnv(platform=None):
   #              header that causes the kernel to set the READ_IMPLIES_EXEC
   #              personality flag, which disables NX page protection.
   linux_env.Prepend(
+      CFLAGS=['-O2'],
+      CXXFLAGS=['-O2'],
       CPPDEFINES=[['-D_FORTIFY_SOURCE', '2']],
-      LINKFLAGS=['-pie', '-Wl,-z,relro', '-Wl,-z,now', '-Wl,-z,noexecstack'],
+      LINKFLAGS=['-O2', '-pie', '-Wl,-z,relro', '-Wl,-z,now', '-Wl,-z,noexecstack'],
       )
   # The ARM toolchain has a linker that doesn't handle the code its
   # compiler generates under -fPIE.
diff --git a/tools/Makefile b/tools/Makefile
index df922ac047476ef944..ab9351c178047ed43b 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -34,16 +34,16 @@ THISMAKEFILE := $(lastword $(MAKEFILE_LIST))
 SHELL = /bin/bash
 
 # By default, checkout from read-only repository:
-#   http://git.chromium.org/native_client
+#   http://chromium.googlesource.com/native_client
 # Maintainers can either override this with read-write repository:
 #   ssh://gerrit.chromium.org/native_client
 # or add to git config:
 #   [url "ssh://gerrit.chromium.org"]
-#     pushinsteadof = http://git.chromium.org
+#     pushinsteadof = http://chromium.googlesource.com
 # or add to git-cl codereview.settings:
 #   PUSH_URL_CONFIG: url.ssh://gerrit.chromium.org.pushinsteadof
-#   ORIGIN_URL_CONFIG: http://git.chromium.org
-GIT_BASE_URL = http://git.chromium.org/native_client
+#   ORIGIN_URL_CONFIG: http://chromium.googlesource.com
+GIT_BASE_URL = http://chromium.googlesource.com/native_client
 
 CROSSARCH = x86_64-nacl
 TOOLCHAINLOC ?= out
@@ -131,11 +131,11 @@ endif
 ##################################################################
 #  The version numbers for the tools we will be building.
 ##################################################################
-GMP_VERSION = 5.0.2
-MPFR_VERSION = 3.0.1
-MPC_VERSION = 0.9
-PPL_VERSION = 0.11.2
-CLOOG_PPL_VERSION = 0.15.9
+GMP_VERSION = 6.1.0
+MPFR_VERSION = 3.1.4
+MPC_VERSION = 1.0.3
+PPL_VERSION = 1.2pre4
+CLOOG_PPL_VERSION = 0.15.11
 define GCC_EXTRAS
 gmp: $(GMP_VERSION), \
 mpfr: $(MPFR_VERSION), \
@@ -202,22 +202,28 @@ sync: | $(all-latest-git-sources)
 # They are only updated when pre-requisites version changes.
 .PHONY: gcc-extras
 gcc-extras:
-	if [[ "$$(cat BUILD/.gcc-extras-version)" != "$(GCC_EXTRAS)" ]]; then \
-	    rm -rf BUILD/.gcc-extra-* SRC/.gcc-extra-* && \
-	    $(MAKE) -f $(THISMAKEFILE) install-gcc-extras && \
-	    echo -n "$(GCC_EXTRAS)" > BUILD/.gcc-extras-version; \
+	if [[ "$$(cat BUILD/.gcc-extras-version)" != "$(GCC_EXTRAS)" ]] ; then \
+		rm -rf BUILD/.gcc-extra-* SRC/.gcc-extra-* && \
+		$(MAKE) -f $(THISMAKEFILE) install-gcc-extras && \
+		echo -n "$(GCC_EXTRAS)" > BUILD/.gcc-extras-version; \
 	fi
 
-gcc-extras := gmp mpfr mpc ppl cloog-ppl
+# TODO: fix ppl linking errors
+# gcc-extras := gmp mpfr mpc ppl cloog-ppl
+gcc-extras := gmp mpfr mpc ppl
 
 gcc_extra_dir = $(subst cloog-ppl,cloog,$*)
 gcc_extra_version = $($(shell echo $*| tr '[:lower:]-' '[:upper:]_')_VERSION)
 
 all-src-gcc-extras = $(gcc-extras:%=SRC/.gcc-extra-%)
+
 $(all-src-gcc-extras): SRC/.gcc-extra-%: | SRC
 	rm -rf SRC/.gcc-extra-$*
-	cd SRC && tar xpf $(SRCDIR)/$(gcc_extra_dir)/$*-$(gcc_extra_version).tar.*
-	mv SRC/$*-$(gcc_extra_version) SRC/.gcc-extra-$*
+	cd SRC && \
+		tar xpf $(SRCDIR)/$(gcc_extra_dir)/$*-$(gcc_extra_version).tar.*
+	mv \
+		SRC/$*-$(gcc_extra_version) \
+		SRC/.gcc-extra-$*
 
 # All Macs need Core2 assembly and --enable-fat is broken with stock MacOS gcc.
 ifneq ($(PLATFORM), mac)
@@ -242,53 +248,83 @@ endef
 
 BUILD/.gcc-extra-build-gmp: | BUILD SRC/.gcc-extra-gmp
 	$(gcc-extra-configure) \
-	    --enable-cxx $(gmp_use_fat_binary) \
-	    ABI="$(HOST_TOOLCHAIN_BITS)"
+		--enable-cxx \
+		$(gmp_use_fat_binary) \
+		ABI="$(HOST_TOOLCHAIN_BITS)"
 
-BUILD/.gcc-extra-build-mpfr: | SRC/.gcc-extra-mpfr BUILD/.gcc-extra-install-gmp
-	$(gcc-extra-configure) --with-gmp=$$BUILD/.gcc-extra-install-gmp
+BUILD/.gcc-extra-build-mpfr: | SRC/.gcc-extra-mpfr \
+					BUILD/.gcc-extra-install-gmp
+	CFLAGS=-I$$BUILD/.gcc-extra-build-gmp/include \
+		CXXFLAGS=-I$$BUILD/.gcc-extra-build-gmp/include \
+		LDFLAGS=-L$$BUILD/.gcc-extra-build-gmp/lib \
+		$(gcc-extra-configure) \
+			--with-gmp=$$BUILD/.gcc-extra-install-gmp
 
-BUILD/.gcc-extra-build-mpc: | SRC/.gcc-extra-mpc BUILD/.gcc-extra-install-gmp \
-                                                 BUILD/.gcc-extra-install-mpfr
-	$(gcc-extra-configure) \
-	    --with-gmp=$$BUILD/.gcc-extra-install-gmp \
-	    --with-mpfr=$$BUILD/.gcc-extra-install-mpfr
+BUILD/.gcc-extra-build-mpc: | SRC/.gcc-extra-mpc \
+					BUILD/.gcc-extra-install-gmp \
+					BUILD/.gcc-extra-install-mpfr
+	CFLAGS=-I$$BUILD/.gcc-extra-build-gmp/include \
+		CXXFLAGS=-I$$BUILD/.gcc-extra-build-gmp/include \
+		LDFLAGS=-L$$BUILD/.gcc-extra-build-gmp/lib \
+		$(gcc-extra-configure) \
+			--with-gmp=$$BUILD/.gcc-extra-install-gmp \
+			--with-mpfr=$$BUILD/.gcc-extra-install-mpfr
 
-BUILD/.gcc-extra-build-ppl: | SRC/.gcc-extra-ppl BUILD/.gcc-extra-install-gmp
-	$(gcc-extra-configure) \
-	    --enable-interfaces="cxx c" \
-	    --with-gmp-prefix=$$BUILD/.gcc-extra-install-gmp
+BUILD/.gcc-extra-build-ppl: | SRC/.gcc-extra-ppl \
+					BUILD/.gcc-extra-install-gmp
+	CFLAGS=-I$$BUILD/.gcc-extra-build-gmp/include \
+		CXXFLAGS=-I$$BUILD/.gcc-extra-build-gmp/include \
+		LDFLAGS=-L$$BUILD/.gcc-extra-build-gmp/lib \
+		$(gcc-extra-configure) \
+			--enable-interfaces="cxx c" \
+			--with-gmp=$$BUILD/.gcc-extra-install-gmp
+	sed -e 's|demos ||g' \
+		<BUILD/.gcc-extra-build-ppl/Makefile \
+		>BUILD/.gcc-extra-build-ppl/Makefile.out
+	cat \
+		<BUILD/.gcc-extra-build-ppl/Makefile.out \
+		>BUILD/.gcc-extra-build-ppl/Makefile
+	$(RM) BUILD/.gcc-extra-build-ppl/Makefile.out
 
 .PHONY: fix-cloog-ppl-check
 fix-cloog-ppl-check: | SRC/.gcc-extra-cloog-ppl
-	for i in SRC/.gcc-extra-cloog-ppl/configure{.in,} ; do \
-	    sed -e s'|ppl_minor_version=10|ppl_minor_version=11|'<$$i >$$i.out;\
-	    cat $$i.out >$$i; \
-	    rm $$i.out; \
+	for i in SRC/.gcc-extra-cloog-ppl/configure{.in,}; do \
+		sed -e s'|ppl_minor_version=10|ppl_minor_version=11|' \
+			<$$i \
+			>$$i.out;\
+		cat \
+			<$$i.out \
+			>$$i; \
+		rm $$i.out; \
 	done
 	sed -e s'|LIBS = @LIBS@|LIBS = @LIBS@ -lstdc++ -lm |' \
-	    <SRC/.gcc-extra-cloog-ppl/Makefile.in \
-	    >SRC/.gcc-extra-cloog-ppl/Makefile.in.out
-	cat SRC/.gcc-extra-cloog-ppl/Makefile.in.out \
-	    >SRC/.gcc-extra-cloog-ppl/Makefile.in
-	rm SRC/.gcc-extra-cloog-ppl/Makefile.in.out
+		<SRC/.gcc-extra-cloog-ppl/Makefile.in \
+		>SRC/.gcc-extra-cloog-ppl/Makefile.in.out
+	cat \
+		<SRC/.gcc-extra-cloog-ppl/Makefile.in.out \
+		>SRC/.gcc-extra-cloog-ppl/Makefile.in
+	$(RM) SRC/.gcc-extra-cloog-ppl/Makefile.in.out
 	find SRC/.gcc-extra-cloog-ppl -print0 | \
-	    xargs -0 touch -r SRC/.gcc-extra-cloog-ppl
+		xargs -0 touch -r SRC/.gcc-extra-cloog-ppl
 
 BUILD/.gcc-extra-build-cloog-ppl: fix-cloog-ppl-check | \
-                                                    BUILD/.gcc-extra-install-ppl
-	$(gcc-extra-configure) \
-	    --with-gmp=$$BUILD/.gcc-extra-install-gmp \
-	    --with-ppl=$$BUILD/.gcc-extra-install-ppl
+					BUILD/.gcc-extra-install-gmp \
+					BUILD/.gcc-extra-install-ppl
+	CFLAGS=-I$$BUILD/.gcc-extra-build-ppl/include \
+		CXXFLAGS=-I$$BUILD/.gcc-extra-build-ppl/include \
+		LDFLAGS=-L$$BUILD/.gcc-extra-build-ppl/lib \
+		$(gcc-extra-configure) \
+			--with-gmp=$$BUILD/.gcc-extra-install-gmp \
+			--with-ppl=$$BUILD/.gcc-extra-install-ppl
 
 all-install-gcc-extras = $(gcc-extras:%=BUILD/.gcc-extra-install-%)
 $(all-install-gcc-extras): BUILD/.gcc-extra-install-%: | SRC \
-                                                        BUILD/.gcc-extra-build-%
+					BUILD/.gcc-extra-build-%
 	if mkdir BUILD/.gcc-extra-install-$*; then \
-	    mkdir BUILD/.gcc-extra-install-$*/include && \
-	    mkdir BUILD/.gcc-extra-install-$*/lib && \
-	    ln -s lib BUILD/.gcc-extra-install-$*/lib64 && \
-	    ln -s lib BUILD/.gcc-extra-install-$*/lib32; \
+		mkdir BUILD/.gcc-extra-install-$*/include && \
+		mkdir BUILD/.gcc-extra-install-$*/lib && \
+		ln -s lib BUILD/.gcc-extra-install-$*/lib64 && \
+		ln -s lib BUILD/.gcc-extra-install-$*/lib32; \
 	fi
 	cd BUILD/.gcc-extra-build-$* && $(MAKE) install
 
@@ -364,7 +400,7 @@ binutils: BUILD/stamp-$(CROSSARCH)-binutils
 # Builds the cross gcc used to build the libraries.
 ##################################################################
 
-GCC_SRC_DIR = $(abspath SRC/gcc)
+GCC_SRC_DIR = $(abspath SRC/gcc/src)
 
 GMP_DIR = $(abspath BUILD/.gcc-extra-install-gmp)
 MPFR_DIR = $(abspath BUILD/.gcc-extra-install-mpfr)
@@ -380,31 +416,32 @@ else
 lstdc++ = -lstdc++
 endif
 
+# TODO: fix cloog build
 ifneq ($(SRCDIR),)
 GCC_EXTRAS_FLAGS = \
-    --with-gmp=$(GMP_DIR) \
-    --with-mpfr=$(MPFR_DIR) \
-    --with-ppl=$(PPL_DIR) \
-    --with-host-libstdcxx="-lpwl $(lstdc++) -lm" \
-    --with-cloog=$(CLOOG_DIR) \
-    --disable-ppl-version-check
+	--with-gmp=$(GMP_DIR) \
+	--with-mpfr=$(MPFR_DIR) \
+	--with-ppl=$(PPL_DIR) \
+	--with-host-libstdcxx="-lpwl $(lstdc++) -lm" \
+	--disable-ppl-version-check
+	# --with-cloog=$(CLOOG_DIR)
 else
 GCC_EXTRAS_FLAGS = \
-    --with-gmp \
-    --with-mpfr \
-    --with-ppl \
-    --with-cloog
+	--with-gmp \
+	--with-mpfr \
+	--with-ppl
+	# --with-cloog
 endif
 
 GCC_CONFIGURE_FLAGS = \
-    --disable-decimal-float \
-    --disable-libgomp \
-    --disable-libmudflap \
-    --disable-libssp \
-    --disable-libstdcxx-pch \
-    --target=$(CROSSARCH) \
-    --enable-linker-build-id \
-    $(GCC_EXTRAS_FLAGS)
+	--disable-decimal-float \
+	--disable-libgomp \
+	--disable-libmudflap \
+	--disable-libssp \
+	--disable-libstdcxx-pch \
+	--target=$(CROSSARCH) \
+	--enable-linker-build-id \
+	$(GCC_EXTRAS_FLAGS)
 
 ifdef MULTILIB
 ifeq ($(MULTILIB),no)
@@ -423,8 +460,8 @@ GCC_CXX = g++ -m$(HOST_TOOLCHAIN_BITS)
 endif
 
 GCC_DEFINES = \
-    -Dinhibit_libc \
-    -D__gthr_posix_h
+	-Dinhibit_libc \
+	-D__gthr_posix_h
 
 
 GCC_CFLAGS_FOR_TARGET-nolibc =
@@ -890,7 +927,7 @@ INST_GLIBC_PROGRAM ?= none
 .PHONY: build-with-glibc
 build-with-glibc: SRC/gcc
 	$(MAKE) -f $(THISMAKEFILE) sdkdirs
-	cp -f SRC/gcc/COPYING* $(DESTDIR)$(PREFIX)
+	cp -f SRC/gcc/src/COPYING* $(DESTDIR)$(PREFIX)
 	$(MAKE) -f $(THISMAKEFILE) BUILD/stamp-$(CROSSARCH)-binutils
 ifeq ($(INST_GLIBC_PROGRAM), none)
 	$(MAKE) -f $(THISMAKEFILE) BUILD/stamp-$(CROSSARCH)-pregcc-standalone
@@ -954,7 +991,7 @@ endif
 .PHONY: build-with-newlib
 build-with-newlib: SRC/gcc
 	$(MAKE) -f $(THISMAKEFILE) sdkdirs
-	cp -f SRC/gcc/COPYING* $(DESTDIR)$(PREFIX)
+	cp -f SRC/gcc/src/COPYING* $(DESTDIR)$(PREFIX)
 	$(MAKE) -f $(THISMAKEFILE) BUILD/stamp-$(CROSSARCH)-binutils
 	$(MAKE) -f $(THISMAKEFILE) BUILD/stamp-$(CROSSARCH)-pregcc
 	$(MAKE) -f $(THISMAKEFILE) export-headers
@@ -986,7 +1023,7 @@ buildbot-build-with-newlib: | \
 # because these are not changed from upstream.
 BINUTILS_PATCHNAME := naclbinutils-$(BINUTILS_VERSION)-r$(shell $(SVNVERSION) | tr : _)
 GCC_PATCHNAME := naclgcc-$(GCC_VERSION)-r$(shell $(SVNVERSION) | tr : _)
-#GDB_PATCHNAME := naclgdb-$(GDB_VERSION)-r$(shell $(SVNVERSION) | tr : _)
+GDB_PATCHNAME := naclgdb-$(GDB_VERSION)-r$(shell $(SVNVERSION) | tr : _)
 GLIBC_PATCHNAME := naclglibc-$(GLIBC_VERSION)-r$(shell $(SVNVERSION) | tr : _)
 NEWLIB_PATCHNAME := naclnewlib-$(NEWLIB_VERSION)-r$(shell $(SVNVERSION) | tr : _)
 
